crit/mempages.go

```diff
@@ -193,3 +195,96 @@ func (mr *MemoryReader) GetShmemSize() (int64, error) {
 
 	return size, nil
 }
+
+// PatternMatch represents a match when searching for a pattern in memory.
+type PatternMatch struct {
+	Vaddr   uint64
+	Length  int
+	Context int
+	Match   string
+}
+
+// SearchPattern searches for a pattern in the process memory pages.
+func (mr *MemoryReader) SearchPattern(pattern string, context, chunkSize int) ([]PatternMatch, error) {
+	if context < 0 {
+		return nil, errors.New("context size cannot be negative")
+	}
+
+	// Set a default chunk size of 10MB to be read at a time
+	if chunkSize <= 0 {
+		chunkSize = 10 * 1024 * 1024
+	}
+
+	regexPattern, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []PatternMatch
+
+	f, err := os.Open(filepath.Join(mr.checkpointDir, fmt.Sprintf("pages-%d.img", mr.pagesID)))
```

Comment:
When testing the changes in the current PR for checkpointctl (with a checkpoint crated with Podman), the search functionality fails with the following error:
```
$ sudo ./checkpointctl memparse -s test /tmp/chkpt.tar.gz 
Error: failed to create memory reader: error opening binary file: open /tmp/checkpointctl4159436349/checkpoint/pagemap-0.img: no such file or directory

$ sudo tar tvf /tmp/chkpt.tar.gz | grep pagemap
-rw-r--r-- root/root       254 2024-07-25 17:24 checkpoint/pagemap-1.img
```

Reply:
Oh, I see where this comes from.  This is due to the fact that the search is done per PID, so a PID need to be provided. I will add a more meaningful error message, or should we allow the search to go through all PIDs?

Reply:
> Should we check that the value of `mr.pagesID` is >= 1 here?

Yes, I will add a check for valid PID.

Reply:
> should we allow the search to go through all PIDs?

It would make sense to support search through all PIDs with something like `-a, --all` option.

Reply:
I would suggest that for now, we can check if the PID > 0 and allow searching only for a single PID. We can add the feature to search across all PIDs in a separate PR.

Reply:
Wouldn't it be better check that PID is valid as we did [here](https://github.com/checkpoint-restore/checkpointctl/blob/723e647f0a1fe188ee6a3953570843aa546802a5/cmd/memparse.go#L172)  so that we can output a more meaningful error message? Checking if PID > 0 won't guarantee that the PID is valid, and the user will receive the same error message otherwise.

Reply:
> Searching through all PIDs would be a little more involved since we'd have to query the process tree and fetch all PIDs from there

Perhaps it would be better to keep the functionality in go-criu simple and use something like a loop in checkpointctl that iterates over all PIDs.